src/dune_engine/build_context.ml

```diff
@@ -6,6 +6,9 @@ type t =
   ; build_dir : Path.Build.t
   ; env : Env.t
   ; host : t option
+  ; stdlib_dir : Path.t  (** todo try to remove it *)
```

Comment:
@cwong-ocaml @aalekseyev this change is unfortunate, but it seems inevitable if artifact substitution is part of the engine.

Reply:
I think it might be a good idea to see if we can rewrite `Install` in terms of a lower-level filesystem-as-known-to-dune API (I can also work on putting something like that together); I'm sure this isn't the only module that would benefit from it.

Reply:
My hope would be that artifact substitution could be a part of the rules, instead. (I don't think we should think too much about that for this PR, though)